crates/bevy_ecs/src/component.rs

```diff
@@ -336,57 +408,68 @@ impl ComponentDescriptor {
             name: Cow::Borrowed(std::any::type_name::<T>()),
-            storage_type,
+            storage_type: StorageType::Table,
             is_send_and_sync: false,
             type_id: Some(TypeId::of::<T>()),
             layout: Layout::new::<T>(),
             drop: needs_drop::<T>().then_some(Self::drop_ptr::<T> as _),
         }
     }
-
-    #[inline]
-    pub fn storage_type(&self) -> StorageType {
-        self.storage_type
-    }
-
-    #[inline]
-    pub fn type_id(&self) -> Option<TypeId> {
-        self.type_id
-    }
-
-    #[inline]
-    pub fn name(&self) -> &str {
-        self.name.as_ref()
-    }
 }
 
-#[derive(Debug, Default)]
+/// The metadata store for all components within a [`World`].
+///
+/// A public read-only API is accessbile. To initialize a new component,
+/// use [`World::init_component`], [`World::init_resource`], [`World::init_non_send_resource`] or
+/// [`World::init_component_with_descriptor`].  Initialized components cannot be removed or
+/// altered from a world.
+///
+/// [`World`]: crate::world::World
+/// [`World::init_component`]: crate::world::World::init_component
+/// [`World::init_resource`]: crate::world::World::init_resource
+/// [`World::init_non_send_resource`]: crate::world::World::init_non_send_resource
+/// [`World::init_component_with_descriptor`]: crate::world::World::init_component_with_descriptor
+#[derive(Debug)]
 pub struct Components {
     components: Vec<ComponentInfo>,
-    indices: std::collections::HashMap<TypeId, usize, fxhash::FxBuildHasher>,
-    resource_indices: std::collections::HashMap<TypeId, usize, fxhash::FxBuildHasher>,
+    ids: HashMap<TypeId, ComponentId, fxhash::FxBuildHasher>,
+    resource_ids: HashMap<TypeId, ComponentId, fxhash::FxBuildHasher>,
 }
 
 impl Components {
+    pub(crate) fn new() -> Components {
+        Components {
+            components: Vec::new(),
+            ids: HashMap::default(),
+            resource_ids: HashMap::default(),
+        }
+    }
+
     #[inline]
-    pub fn init_component<T: Component>(&mut self, storages: &mut Storages) -> ComponentId {
+    pub(crate) fn init_component<T: Component>(&mut self, storages: &mut Storages) -> ComponentId {
         let type_id = TypeId::of::<T>();
 
         let Components {
-            indices,
-            components,
-            ..
+            ids, components, ..
         } = self;
-        let index = indices.entry(type_id).or_insert_with(|| {
-            Components::init_component_inner(components, storages, ComponentDescriptor::new::<T>())
+        let id = ids.entry(type_id).or_insert_with(|| {
+            let index = Components::init_component_inner(
+                components,
+                storages,
+                ComponentDescriptor::component_from_type::<T>(),
+            );
+            ComponentId(index)
         });
-        ComponentId(*index)
+        *id
     }
 
-    pub fn init_component_with_descriptor(
+    pub(crate) fn init_component_with_descriptor(
```

Comment:
This function is extremely useful for scripting when you want to store references to script objects as components of an entity.

Because all script objects will be hidden behind one type `T`, it is necessary to register multiple `ComponentIds` for one rust type `T`, which is exactly what can be done using `init_component_with_descriptor`.

Reply:
As it currently stands, this API is not publicly usable. `World::components_mut` does not exist, so unless a user is constructing their own `Components` via `Default` independently of `World` and editing it. `World::init_component_descriptor` is the publicly usable version of this function.

Reply:
Ah, thats true!